### PR TITLE
Update `CsrfTokenApi` to retry failed requests due to invalid tokens

### LINKED ISSUE #12372: Retry UI requests that fail because of CSRF token validation.

### PR DESCRIPTION
**Pull request by bunchesofdonald**
During testing I noticed that occasionally a request would be denied due to an invalid CSRF token. This seems to be the result of a race condition where the token is able to be refreshed while the change request is in-flight. To solve for this I updated the the `CsrfTokenApi` to retry failed requests, a single time, when they're due to invalid tokens.

Closes #12372

### CODE REVIEW

File: ui/src/services/csrfTokenApi.ts [hunk 1]
```diff
@@ -131,4 +137,21 @@ export function setupCsrfInterceptor(csrfTokenApi: CreateActions<CsrfTokenApi>,
 
         return config
     })
+
+    axiosInstance.interceptors.response.use(undefined, async (error: AxiosError) => {
```
Comment (znicholasbrown) [inline ui/src/services/csrfTokenApi.ts:141]: What does passing `undefined` here and a function as the second arg do? Is the first for handling success and the second for errors?

File: ui/src/services/csrfTokenApi.ts [hunk 2]
```diff
@@ -131,4 +137,21 @@ export function setupCsrfInterceptor(csrfTokenApi: CreateActions<CsrfTokenApi>,
...
+        if (isAxiosError(error) && csrfTokenApi.isInvalidCsrfToken(error)) {
+            const config = error.config
+
+            if (config && config.headers['Prefect-Csrf-Retry-Count']) {
+                const retryCount = parseInt(config.headers['Prefect-Csrf-Retry-Count'], 10)
+                if (retryCount < MAX_RETRIES) {
+                    await csrfTokenApi.addCsrfHeaders(config)
+                    config.headers['Prefect-Csrf-Retry-Count'] = (retryCount + 1).toString()
```
Comment (znicholasbrown) [inline ui/src/services/csrfTokenApi.ts:149]: Just double checking - this header is echoed back by the API, right? Or is the intent that if this fails a second time it'll have the same axios instance and so the header will be present from config (assuming this has run previously)

**Review by znicholasbrown — APPROVED** [review]
A couple of questions, mostly for my own clarification

File: ui/src/services/csrfTokenApi.ts (hunk 1 above)
Comment (bunchesofdonald) [reply ui/src/services/csrfTokenApi.ts:141]: Yep, exactly, the first handle success and second handles errors.

File: ui/src/services/csrfTokenApi.ts (hunk 2 above)
Comment (bunchesofdonald) [reply ui/src/services/csrfTokenApi.ts:149]: It's no echoed back by the API. The `config` here is the request we initially fired that resulted in the error. So we're essentially just using the headers as data storage.

Comment (znicholasbrown) [reply ui/src/services/csrfTokenApi.ts:149]: Ok got it, so it's the second - that makes more sense, thank you

**Pull request merged**